flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java

```diff
@@ -66,33 +91,47 @@ public void open(
                 this.checkpointedState == null,
                 "The " + getClass().getSimpleName() + " has already been initialized.");
 
-        this.checkpointedState =
-                context.getOperatorStateStore()
-                        .getListState(
-                                new ListStateDescriptor<>(
-                                        name + "-sequence-state", LongSerializer.INSTANCE));
-        this.valuesToEmit = new ArrayDeque<>();
-        if (context.isRestored()) {
-            // upon restoring
+        final ListStateDescriptor<SubTaskState> stateDescriptor =
+                new ListStateDescriptor<>(
+                        name + "-sequence-state", TypeInformation.of(SubTaskState.class));
+        this.checkpointedState = context.getOperatorStateStore().getListState(stateDescriptor);
 
-            for (Long v : this.checkpointedState.get()) {
-                this.valuesToEmit.add(v);
-            }
+        if (!subTaskStates.isEmpty()) {
+            this.subTaskStates.clear();
+        }
```

Comment:
```suggestion
        this.subTaskStates.clear();
```
nit: clearing an empty collection shouldn't hurt.